spark3/src/test/java/org/apache/iceberg/actions/TestRemoveFilesAction.java

```diff
@@ -0,0 +1,317 @@
+      .withPartition(TestHelpers.Row.of(2))
+      .withRecordCount(1)
+      .build();
+  static final DataFile FILE_D = DataFiles.builder(SPEC)
+      .withPath("/path/to/data-d.parquet")
+      .withFileSizeInBytes(10)
+      .withPartition(TestHelpers.Row.of(3))
+      .withRecordCount(1)
+      .build();
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  private Table table;
+
+  @Before
+  public void setupTableLocation() throws Exception {
+    File tableDir = temp.newFolder();
+    String tableLocation = tableDir.toURI().toString();
+    this.table = TABLES.create(SCHEMA, SPEC, Maps.newHashMap(), tableLocation);
+    spark.conf().set("spark.sql.shuffle.partitions", SHUFFLE_PARTITIONS);
+  }
+
+  private void checkRemoveFilesResults(long expectedDatafiles, long expectedManifestsDeleted,
+                                       long expectedManifestListsDeleted, long expectedOtherFilesDeleted,
+                                       RemoveFiles.Result results) {
+    Assert.assertEquals("Incorrect number of manifest files deleted",
+        expectedManifestsDeleted,  results.removedManifestsCount());
+    Assert.assertEquals("Incorrect number of datafiles deleted",
+        expectedDatafiles, results.removedDataFilesCount());
+    Assert.assertEquals("Incorrect number of manifest lists deleted",
+        expectedManifestListsDeleted, results.removedManifestListsCount());
+    Assert.assertEquals("Incorrect number of other lists deleted",
+        expectedOtherFilesDeleted - 1, results.otherRemovedFilesCount());
+  }
+
+  @Test
+  public void dataFilesCleanupWithParallelTasks() {
+    table.newFastAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    table.newFastAppend()
+        .appendFile(FILE_B)
+        .commit();
+
+    table.newRewrite()
+        .rewriteFiles(ImmutableSet.of(FILE_B), ImmutableSet.of(FILE_D))
+        .commit();
+
+    table.newRewrite()
+        .rewriteFiles(ImmutableSet.of(FILE_A), ImmutableSet.of(FILE_C))
+        .commit();
+
+    Set<String> deletedFiles = Sets.newHashSet();
+    Set<String> deleteThreads = ConcurrentHashMap.newKeySet();
+    AtomicInteger deleteThreadsIndex = new AtomicInteger(0);
+
+    RemoveFiles.Result result = getSparkActions().removeFiles(getMetadataLocation(table))
+        .io(table.io())
+        .executeRemoveWith(Executors.newFixedThreadPool(4, runnable -> {
+          Thread thread = new Thread(runnable);
+          thread.setName("remove-files-" + deleteThreadsIndex.getAndIncrement());
+          thread.setDaemon(true); // daemon threads will be terminated abruptly when the JVM exits
+          return thread;
+        }))
+        .removeWith(s -> {
+          deleteThreads.add(Thread.currentThread().getName());
+          deletedFiles.add(s);
+        })
+        .execute();
+
+    // Verifies that the delete methods ran in the threads created by the provided ExecutorService ThreadFactory
+    Assert.assertEquals(deleteThreads,
```

Comment:
Is it guaranteed all 4 threads are going to be used?

Reply:
checked org.apache.iceberg.util.Tasks.Builder#runParallel (The method that uses executor service to schedule the tasks), since the tasks are submitted serially in a loop, all 4 threads should be used

Reply:
Thanks for checking!